src/Service/NixGenerator.php

```diff
@@ -0,0 +1,392 @@
+                            $dst
+                        );
+
+                    if (false === $copy) {
+                        $this
+                            ->io
+                            ->writeError(
+                                '<error>' .
+                                'Preloading into Nix store failed: could' .
+                                ' not write to temporary directory.' .
+                                '</error>'
+                            );
+
+                        return null;
+                    }
+
+                    return $dst;
+                },
+                array_filter(
+                    $collected,
+                    static fn (array $info): bool => 'cache' === $info['type']
+                )
+            )
+        );
+
+        if ([] === $toPreload) {
+            return;
+        }
+
+        try {
+            // Preload in batches, to keep the exec arguments reasonable.
+            $process = new ProcessExecutor($this->io);
+            $numPreloaded = 0;
+
+            foreach (array_chunk($toPreload, 100) as $chunk) {
+                $command = sprintf(
+                    'nix-store --add-fixed sha256 %s',
+                    implode(
+                        ' ',
+                        array_map(['Composer\\Util\\ProcessExecutor', 'escape'], $chunk)
+                    )
+                );
+
+                if ($process->execute($command, $output) !== Command::SUCCESS) {
+                    $this->io->writeError('<error>Preloading into Nix store failed.</error>');
+                    $this->io->writeError($output);
+
+                    break;
+                }
+
+                $numPreloaded += count($chunk);
+            }
+
+            $this
+                ->io
+                ->writeError(
+                    sprintf(
+                        '<info>Preloaded %d packages into the Nix store.</info>',
+                        $numPreloaded
+                    )
+                );
+        } finally {
+            $this->fs->removeDirectory($tempDir);
+        }
+    }
+
+    public function shouldPreload(): bool
+    {
+        return $this->shouldPreload;
+    }
+
+    private function fetch(
+        BasePackage $package,
+        string $name,
+        string $cacheFile
+    ): string {
+        // If some packages were previously installed but since removed from
+        // cache, `sha256` will be false for those packages in `collected`.
+        // Here, we amend cache by refetching, so we can then determine the
+        // file hash again.
+        $downloader = $this->composer->getDownloadManager()->getDownloader('file');
+
+        $tempDir = $this->cacheDir . '.nixify-tmp-' . substr(md5(uniqid('', true)), 0, 8);
+        $this->fs->ensureDirectoryExists($tempDir);
+
+        $this->io->writeError(sprintf(
+            '<info>Nixify could not find cache for package $s, which will be refetched</info>',
+            $name
+        ));
```

Comment:
I feel like the output of this can be quite spammy when the amount of packages to fetch is large. Hence why I tallied them before fetching.

I do like the generator changes in `collect` though. Maybe we can make that return a `needsFetch` boolean? And then the caller can be responsible for printing a summary and calling `fetch`.

But I'm open to other ideas. 🙂